### PR TITLE
Add nwb support I

### PR DESCRIPTION
**Pull request by h-mayorquin**
This is the first attempt at importing the nwb IO functionality from the main module. This PR specifically focuses on writing and attemps to keep most of the conventions that you had in the main repo code with some -to be discussed changes.

There are two main functions that are added and tested:
1. `write_labels_to_nwb` that takes as an input an labels object and a path to write the nwb file.
2. `append_labels_data_to_nwb` which takes an input a labels object and an in-memory nwb file. 

The first function uses the second to do most of the heavy work but I added it the two possibilities open. I added tests for  `append_labels_data_to_nwb`  with two example datasets files from your fixtures.

Related discussion can be found here:
https://github.com/talmolab/sleap-io/discussions/10

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -1,4 +1,5 @@
 {
+    "version": "0.2.0",
```
Comment (h-mayorquin) [inline .vscode/launch.json:2]: I just added this line out of curiosity. My testing with vscode was crashing without this (I could not call the debugger). Is there a reason why did you remove this from the default?

File: sleap_io/io/nwb.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        description=container_description,
+        original_videos=[f"{video.filename}"],
+        labeled_videos=[f"{video.filename}"],
+        source_software="SLEAP",
+        nodes=skeleton.node_names,
+        edges=np.array(skeleton.edge_inds).astype("uint64"),
+        # dimensions=np.array([[video.backend.height, video.backend.width]]),
+        # scorer=str(labels.provenance),
+        # source_software_version=f"{sleap.__version__}",
+        # To-discuss this in PR.
```
Comment (h-mayorquin) [inline sleap_io/io/nwb.py:228]: Is there an example data of yours that has provenance so I can see how this could be written?
I also think that appending the software version directly (that is, the assumption that the writing is taking place in the same computer that the predictions were done) is probably not a good idea. Maybe that's something that should be on provenance as well?

**Review by h-mayorquin — COMMENTED** [review]
Some more specific comments here and there.

File: sleap_io/io/nwb.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
+"""Functions to write and read from the neurodata without borders (NWB) format. 
+"""
+from tkinter import Label
```
Comment (talmo) [inline sleap_io/io/nwb.py:3]: What's this for?

File: sleap_io/io/nwb.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    """Write labels to an nwb file and save it to the nwbfile_path given
+
+    Args:
+        labels (Labels): A general label object
+        nwbfile_path (str): The path where the nwb file is to be written
+    """
+
+    session_description: str = "Processed SLEAP pose data"
+    session_start_time = datetime.datetime.now(datetime.timezone.utc)
+    identifier = "identifier"
```
Comment (talmo) [inline sleap_io/io/nwb.py:97]: Maybe best to have these as keyword args with defaults for flexibility?

File: sleap_io/io/nwb.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+    )
+
+    nwbfile = append_labels_data_to_nwb(labels, nwbfile)
+
+    with NWBHDF5IO(str(nwbfile_path), "w") as io:
+        io.write(nwbfile)
+
+
+def append_labels_data_to_nwb(labels: Labels, nwbfile: NWBFile) -> NWBFile:
+    """
```
Comment (talmo) [inline sleap_io/io/nwb.py:112]: Missing 1 line description of method

File: sleap_io/io/nwb.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+    Args:
+        processing_module_name (str): The name of the processing module.
+        nwbfile (NWBFile): The nwbfile to attach the processing module to.
+
+    Returns:
+        ProcessingModule: An nwb processing module with the desired name.
+    """
+    if processing_module_name in nwbfile.processing:
+        nwb_processing_module = nwbfile.processing[processing_module_name]
```
Comment (talmo) [inline sleap_io/io/nwb.py:164]: So this is for appending to an existing file? We could get coverage here by saving out one file and then saving out to it again under a different `processing_module_name`

File: sleap_io/io/nwb.py (hunk 2 above)
Comment (talmo) [reply sleap_io/io/nwb.py:228]: Right, these are usually stored in the provenance metadata, but I think the dataset we're currently using as a fixture here might be a bit too old from before we saved that together with the predictions. We can re-generate it using a newer SLEAP version for this PR.

File: sleap_io/io/nwb.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        # To-discuss this in PR.
+    )
+
+    return pose_estimation_container
+
+
+def build_track_pose_estimation_list(
+    track_data_df: pd.DataFrame,
+) -> List[PoseEstimationSeries]:
+    """An auxiliar function to build a list of PoseEstimationSeries associated with
```
Comment (talmo) [inline sleap_io/io/nwb.py:237]: Auxiliary?

File: tests/io/test_nwb.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    with pytest.raises(
+        ValueError, match="No predicted instances found in labels object"
+    ):
+        nwbfile = append_labels_data_to_nwb(labels, nwbfile)
+
+
+def test_typical_case_write(slp_typical):
+    labels = load_slp(slp_typical)
+
+    nwbfile_path = Path(mkdtemp()) / "write_to_nwb_typical_case.nwb"
```
Comment (talmo) [inline tests/io/test_nwb.py:108]: It's generally preferred that you use the `tmp_path` argument instead of managing temporary directories yourself in pytest.

Ref: [Docs](https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html)

**Review by talmo — CHANGES_REQUESTED** [review]

File: .vscode/launch.json (hunk 1 above)
Comment (talmo) [reply .vscode/launch.json:2]: Oh interesting -- I didn't edit it at all, this is what my vscode setup generated. Maybe we shouldn't even have these files checked into the repo at all?

Comment (h-mayorquin) [reply .vscode/launch.json:2]: Yep, they probably should not be there.

File: sleap_io/io/nwb.py (hunk 6 above)
Comment (h-mayorquin) [reply sleap_io/io/nwb.py]: I tried a [ternary operator ](https://docs.python.org/3.10/reference/expressions.html?highlight=ternary#conditional-expressions)approach here in the last commit. We can still add the test if you want but I feel is an overkill.

File: sleap_io/io/nwb.py (hunk 3 above)
Comment (h-mayorquin) [reply sleap_io/io/nwb.py:3]: This got imported because vscode was trying to be too smart. Removed.

File: sleap_io/io/nwb.py (hunk 7 above)
Comment (h-mayorquin) [reply sleap_io/io/nwb.py:238]: Obrigado.

File: sleap_io/io/nwb.py (hunk 5 above)
Comment (h-mayorquin) [reply sleap_io/io/nwb.py:112]: Done

File: sleap_io/io/nwb.py (hunk 4 above)
Comment (h-mayorquin) [reply sleap_io/io/nwb.py:97]: Yep, good idea in case they need to change.

File: tests/io/test_nwb.py (hunk 8 above)
Comment (h-mayorquin) [reply tests/io/test_nwb.py]: Yep. Makes sense.
